dask_cuda/dask_cuda_worker.py

```diff
@@ -98,6 +97,25 @@
     "string (like 5GB or 5000M), "
     "'auto', or zero for no memory management",
 )
+@click.option(
+    "--device-memory-limit",
+    default="auto",
+    help="Bytes of memory per CUDA device that the worker can use. "
+    "This can be an integer (bytes), "
+    "float (fraction of total device memory), "
+    "string (like 5GB or 5000M), "
+    "'auto', or zero for no memory management. "
+    "Note: this parameter is ignored when "
+    "--worker-class is not CUDAWorker.",
+)
+@click.option(
+    "--worker-class",
+    type=str,
+    default="CUDAWorker",
+    help="The Worker class to be used. "
+    "Choosing a non-default worker may result in limited functionality, "
+    "such as no device memory spilling support.",
+)
```

Comment:
For context, the `worker_class` keyword exists because different groups have their own custom Worker classes that they use.  Exposing a keyword like this but making no move to allow them to plug in seems against the spirit of the keyword.

Reply:
> For things like the memory monitor, this may not be an issue, both because we have disabled it temporarily, and because it may just not be necessary with the kind of data used on GPUs.

That's not true. We solely disabled pausing the worker, to control spilling from the device, we still need to monitor the device memory. And this is why I needed to subclass it.

> I would prefer to avoid this until someone asks for it if possible. I'm somewhat against an API like this because it forces us to enumerate the possible options in code. If we were to do something like this I think that we would probably provide the full namespace of the class and then try to import it.

Unfortunately, this is necessary for us to reenable the old `CUDA_VISIBLE_DEVICES` test. Using `CUDAWorker`, which was hardcoded before, prevents us from launching a process on a single-GPU machine to mock test as if it had several and check for the ordering of GPUs of each process.

Reply:
>> For things like the memory monitor, this may not be an issue, both because we have disabled it temporarily, and because it may just not be necessary with the kind of data used on GPUs.

> That's not true. We solely disabled pausing the worker, to control spilling from the device, we still need to monitor the device memory. And this is why I needed to subclass it.

Why do we need to monitor device memory externally from the use of `sizeof`?  If objects' reliably tell us how much device memory they take up then we may not need to track device memory separately.  

Dask workers operated this way for a long time before we finally needed to give in and have a separate periodic callback that tracked system memory.  I'm inclined to try the simpler approach first and see if it breaks down or not.

Reply:
Sure, they do track how much memory they take. However, tracking the device memory lets us decide _when_ it's time to spill memory. Isn't that what `memory_target_fraction` and `memory_spill_fraction` (prepend `device_` for the cases in this PR) are for?

The block https://github.com/rapidsai/dask-cuda/pull/35/files#diff-a77f0c6f19d8d34d59aede5e31455719R282 controls the spilling, and this is why we needed to subclass `Worker`.

Reply:
> Dask workers operated this way for a long time before we finally needed to give in and have a separate periodic callback that tracked system memory. I'm inclined to try the simpler approach first and see if it breaks down or not.

Also, I honestly don't know how they operated before, I don't think I've seen that code (perhaps it doesn't exist in the new codebase anymore?). But if this was necessary to CPU Dask workers, what makes you think it won't be needed here?

I would in general prefer to keep the minimum code as possible (and avoided adding unnecessary code wherever possible), but throwing what we've spent a lot of time developing into the limbo, in particular for something that seems naturally useful, seems like a waste of everyone's time.

Reply:
There are two mechanisms that control spilling:

1.  The `zict.LRU/Buffer` logic, which evicts old data as new data comes in, respecting a maximum limit of bytes.  Historically, this is what Dask workers did first.  It's still active.  It's responsive and pretty simple, but error prone if our estimates of byte sizes are off, or if other processes are taking up memory without our knowledge.

2.  The periodic callback that tracks memory through system utilities.  This is more robust (we don't estimate anything) but involves a lot more code, and in our situation creates the need for more classes and such.

Reply:
> I would in general prefer to keep the minimum code as possible (and avoided adding unnecessary code wherever possible), but throwing what we've spent a lot of time developing into the limbo, in particular for something that seems naturally useful, seems like a waste of everyone's time.

I'm not planning to throw it away.  The work is nicely separable and so shouldn't go stale.

> in particular for something that seems naturally useful

If we find that it's useful then lets merge it in.  I'm inclined to wait until that's clear though.

If the extra features end up not being necessary then you will have my apologies.  I should have been active reviewing earlier to avoid the extra work.

Reply:
> * The `zict.LRU/Buffer` logic, which evicts old data as new data comes in, respecting a maximum limit of bytes.  Historically, this is what Dask workers did first.  It's still active.  It's responsive and pretty simple, but error prone if our estimates of byte sizes are off, or if other processes are taking up memory without our knowledge.
> *

Alright, this was really stupid of me, I knew that, but sometimes is easy to forget the obvious.

But could you tell me what's the reason for tracking memory on `Worker` instead of just relying on the LRU?

Reply:
Nevertheless, I see what you mean now. Given that we can rely on LRU alone, let's try that. I can come up with a few tests quickly tomorrow, and I started https://github.com/rapidsai/dask-cuda/pull/51 for that.

I guess what you expect to see on the `Worker` side of things, that would be to have only something like this: https://github.com/rapidsai/dask-cuda/compare/branch-0.8...pentschev:device-memory-spill-lru?expand=1#diff-34992fbdc24aa7b93a1d01f96d5854af

Is that correct?

Reply:
> But could you tell me what's the reason for tracking memory on Worker instead of just relying on the LRU?

Our `sizeof` estimates are wrong on text and JSON-like data.

Also sometimes other processes take up more memory than expected, or there are memory leaks, or other random things.  

My *hope* is that these issues don't come up as frequently in GPU code.

Reply:
Yes, something like that diff.  You'll also want to add the `data=` keyword to `LocalCUDACluster`.

Reply:
Thanks for the explanation on memory monitor.

> Yes, something like that diff. You'll also want to add the `data=` keyword to `LocalCUDACluster`.

Yes, and I also want to create the object before. :)

But ok, I can probably have it quickly done by tomorrow. There's a few more things that need to be ported to allow it to work (like finding out how much memory the device has in total), and also some test(s), which shouldn't be too difficult now that there's already one that works with the monitoring mechanism and I can base it on that.

Reply:
> Yes, and I also want to create the object before. :)

It's actually valid to pass just the class.  Dask will construct it.  I *think* that this is explained in the `Worker` docstring.  This is better because you're using a Nanny and don't want to pass it through a process boundary.

> like finding out how much memory the device has in total)

I recommend that we start with just using the full memory or a config value by default and not mess with any user inputs (which will get messy).

Reply:
> It's actually valid to pass just the class. Dask will construct it. I _think_ that this is explained in the `Worker` docstring. This is better because you're using a Nanny and don't want to pass it through a process boundary.

Ok, I'll check that.

> I recommend that we start with just using the full memory or a config value by default and not mess with any user inputs (which will get messy).

We need to identify how much memory there is available for the device, regardless. I can probably use the same numba code from before.

Reply:
Right, mostly I want to say lets not add a new `device_memory_foo=` keyword yet if we can avoid it.